gammapy/maps/wcsnd.py

```diff
@@ -688,3 +692,68 @@ def cutout(self, position, width, mode="trim"):
         data = self.data[cutout_slices]
 
         return self._init_copy(geom=geom, data=data)
+
+    def sample_position_energy(self, npred_map=None, n_events=None, random_state=0):
+        """Sample position and energy of events.
+
+        Parameters
+        ----------
+        npred_map : `~gammapy.maps.Map`
+            Predicted number of counts map.
+        n_events : int
+            Number of events to sample.
+        random_state : {int, 'random-seed', 'global-rng', `~numpy.random.RandomState`}
+            Defines random number generator initialisation.
+            Passed to `~gammapy.utils.random.get_random_state`.
+
+        Returns
+        -------
+        coords : `~gammapy.maps.MapCoord` object.
+            Sequence of coordinates and energies of the sampled events.
+        """
+
+        self.random_state = get_random_state(random_state)
+        self.npred_map = npred_map
+        self.n_events = n_events
+
+        sampler = InverseCDFSampler(self.npred_map.data, random_state=self.random_state)
+
+        coords_pix = sampler.sample(n_events)
+        coords = self.npred_map.geom.pix_to_coord(coords_pix[::-1])
+
+        # TODO: pix_to_coord should return a MapCoord object
+        geom = self.npred_map.geom
+        axes_names = ["lon", "lat"] + [ax.name for ax in geom.axes]
+        cdict = OrderedDict(zip(axes_names, coords))
+        cdict["energy"] *= geom.get_axis_by_name("energy").unit
+
+        return MapCoord.create(cdict, coordsys=geom.coordsys)
+
+    def sample_events(self, npred_map=None, n_events=None, random_state=0):
```

Comment:
Please just remove the `sample_events()` method. I think it's easy to create an `astropy.table.Table` object so no need for an extra method.

Reply:
Thanks for the comment @adonath . 
I've changed the code as you indicated. 
About this only point, so do you mean to drop the Table object as output of the `sample_coord()` method and just stop it with:
 
`       ....
        axes_names = ["lon", "lat"] + [ax.name for ax in self.geom.axes]
        cdict = OrderedDict(zip(axes_names, coords))
        cdict["energy"] *= self.geom.get_axis_by_name("energy").unit

        return MapCoord.create(cdict, coordsys=self.geom.coordsys)`
?

Reply:
Yes, I slightly changed my mind here. In the `Map` API we use `MapCoord` objects to handle coordinates and I thought it was more consistent to just have the `sample_coord()` method, which returns a `MapCoord` object. Returning a `Table` object would  not be needed anymore at this point, hence there is no need for the additional `sample_events()` method either. Do you  agree this makes sense?

Reply:
Okay, I think it's reasonable hence it's fine to me to drop the `sample_events()` method. However, I was wondering if we should modify the PIG, in the code example here `https://github.com/fabiopintore/gammapy/blob/event_sampling_pig/docs/development/pigs/pig-009.rst#mapdataseteventsampler` .

Reply:
I would say it's not needed. If we find small improvements during implementation, we just implement those. The PIG outlines the big picture and serves as a basis for discussion. It's not a 100% complete documentation of the development process.